test/subtype.jl

```diff
@@ -0,0 +1,494 @@
+    @test !issub(Tuple{Integer,Integer}, @UnionAll T Tuple{T,T})
+    @test !issub(Tuple{Integer,Int}, (@UnionAll T @UnionAll S<:T Tuple{T,S}))
+    @test !issub(Tuple{Integer,Int}, (@UnionAll T @UnionAll T<:S<:T Tuple{T,S}))
+
+    @test issub_strict((@UnionAll R Tuple{R,R}),
+                       (@UnionAll T @UnionAll S Tuple{T,S}))
+    @test issub_strict((@UnionAll R Tuple{R,R}),
+                       (@UnionAll T @UnionAll S<:T Tuple{T,S}))
+    @test issub((@UnionAll R Tuple{R,R}),
+                (@UnionAll T @UnionAll T<:S<:T Tuple{T,S}))
+    @test issub((@UnionAll R Tuple{R,R}),
+                (@UnionAll T @UnionAll S>:T Tuple{T,S}))
+
+    #@test issub((@UnionAll T @UnionAll T<:S<:T Tuple{T,S}),
+    #            (@UnionAll R Tuple{R,R}))
+
+    @test !issub(Tuple{Real,Real}, @UnionAll T<:Real Tuple{T,T})
+end
+
+# level 3: UnionAll
+function test_3()
+    @test issub_strict(Array{Int,1}, @UnionAll T Vector{T})
+    @test issub_strict((@UnionAll T Pair{T,T}), Pair)
+    @test issub(Pair{Int,Int8}, Pair)
+    @test issub(Pair{Int,Int8}, (@UnionAll S Pair{Int,S}))
+
+    @test !issub((@UnionAll T<:Real T), (@UnionAll T<:Integer T))
+
+    @test isequal_type((@UnionAll T Tuple{T,T}), (@UnionAll R Tuple{R,R}))
+
+    @test !issub((@UnionAll T<:Integer @UnionAll S<:Number Tuple{T,S}),
+                 (@UnionAll T<:Integer @UnionAll S<:Number Tuple{S,T}))
+
+    @test issub_strict((@UnionAll T Tuple{Array{T},Array{T}}),
+                       Tuple{Array, Array})
+
+    AUA = Array{(@UnionAll T Array{T,1}), 1}
+    UAA = (@UnionAll T Array{Array{T,1}, 1})
+
+    @test !issub(AUA, UAA)
+    @test !issub(UAA, AUA)
+    @test !isequal_type(AUA, UAA)
+
+    @test issub_strict((@UnionAll T Int), (@UnionAll T<:Integer Integer))
+
+    @test isequal_type((@UnionAll T @UnionAll S Tuple{T, Tuple{S}}),
+                       (@UnionAll T Tuple{T, @UnionAll S Tuple{S}}))
+
+    @test !issub((@UnionAll T Pair{T,T}), Pair{Int,Int8})
+    @test !issub((@UnionAll T Pair{T,T}), Pair{Int,Int})
+
+    @test isequal_type((@UnionAll T Tuple{T}), Tuple{Any})
+    @test isequal_type((@UnionAll T<:Real Tuple{T}), Tuple{Real})
+
+    @test  issub(Tuple{Array{Integer,1}, Int},
+                 @UnionAll T<:Integer @UnionAll S<:T Tuple{Array{T,1},S})
+
+    @test !issub(Tuple{Array{Integer,1}, Real},
+                 @UnionAll T<:Integer Tuple{Array{T,1},T})
+
+    @test !issub(Tuple{Int,String,Vector{Integer}},
+                 @UnionAll T Tuple{T, T, Array{T,1}})
+    @test !issub(Tuple{String,Int,Vector{Integer}},
+                 @UnionAll T Tuple{T, T, Array{T,1}})
+    @test !issub(Tuple{Int,String,Vector{Tuple{Integer}}},
+                 @UnionAll T Tuple{T,T,Array{Tuple{T},1}})
+
+    @test issub(Tuple{Int,String,Vector{Any}},
+                @UnionAll T Tuple{T, T, Array{T,1}})
+
+    @test isequal_type(Array{Int,1}, Array{(@UnionAll T<:Int T), 1})
+    @test isequal_type(Array{Tuple{Any},1}, Array{(@UnionAll T Tuple{T}), 1})
```

Comment:
I believe it's just that `!(U <: U{T})` is not true in general. Sometimes you can't get a bigger type by wrapping something in a UnionAll.